backend/PIMTool.Core/Helpers/PaginationHelper.cs

```diff
@@ -0,0 +1,74 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using AutoMapper;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Query;
+using PIMTool.Core.Models;
+
+namespace PIMTool.Core.Helpers;
+
+public static class PaginationHelper
+{
+    public static async Task<PaginatedResult> BuildPaginatedResult<T, TDto>(IMapper mapper, IQueryable<T> source, int pageSize, int pageIndex)
+    {
+        var total = await source.CountAsync().ConfigureAwait(false);
+        if (total == 0)
+        {
+            return new PaginatedResult()
+            {
+                PageIndex = 1,
+                PageSize = pageSize,
+                Data = new List<TDto>(),
+                LastPage = 1,
+                IsLastPage = true,
+                Total = total
+            };
+        }
+
+        pageSize = Math.Max(1, pageSize);
+        var lastPage = (int)Math.Ceiling((decimal)total / pageSize);
+        lastPage = lastPage < 1 ? 1 : lastPage;
+        pageIndex = pageIndex > lastPage ? lastPage : pageIndex;
+        var isLastPage = pageIndex == lastPage;
+
+        if (pageIndex > lastPage / 2)
+        {
+            var mod = total % pageSize;
+            var skip = Math.Max((lastPage - pageIndex - 1) * pageSize + mod, 0);
+            var take = isLastPage ? mod : pageSize;
+        
+            var reverse = source.Reverse();
```

Comment:
[C] it was bad performance with reverse here. The code is complicated with skip and take

Reply:
Actually, I used the Reverse() method to improve the performance here. When the page index gets too large (about 100 thousand pages, for 10 items per page), it takes about 10 times slower to load it. So I did a little trick here (use  Reverse() method) to load last pages as fast as those first ones.